src/Microsoft.VisualStudio.ProjectSystem.Managed/Tree/ProjectImportsSubTreeProvider.cs

```diff
@@ -0,0 +1,288 @@
+                                }
+                            }
+                        });
+
+                    return;
+
+                    void SetUpTree()
+                    {
+                        Assumes.Null(_subscription);
+
+                        // Set a visible root
+                        SubmitTreeUpdateAsync(
+                            (currentTree, a, c) =>
+                            {
+                                IProjectTree tree;
+                                if (currentTree == null)
+                                {
+                                    // Create a new tree as no prior one exists
+                                    tree = NewTree(RootNodeCaption, icon: s_rootIcon, flags: ProjectImportsTreeRootFlags);
+                                }
+                                else
+                                {
+                                    // Reuse existing tree, but make it visible
+                                    tree = currentTree.Value.Tree.SetVisible(true);
+                                }
+
+                                return Task.FromResult(new TreeUpdateResult(tree));
+                            });
+
+                        // Subscribe to data to populate the tree and keep it updated with changes
+                        ITargetBlock<IProjectVersionedValue<IProjectImportTreeSnapshot>> actionBlock
+                            = DataflowBlockSlim.CreateActionBlock(
+                                new Action<IProjectVersionedValue<IProjectImportTreeSnapshot>>(SyncTree));
+
+                        IReceivableSourceBlock<IProjectVersionedValue<IProjectImportTreeSnapshot>> importTreeSource
+                            = _projectSubscriptionService.ImportTreeSource.SourceBlock;
+
+                        using (TrySuppressExecutionContextFlow())
+                        {
+                            _subscription = importTreeSource.LinkTo(actionBlock, DataflowOption.PropagateCompletion);
+                        }
+
+                        JoinUpstreamDataSources(_projectSubscriptionService.ImportTreeSource);
+
+                        return;
+
+                        void SyncTree(IProjectVersionedValue<IProjectImportTreeSnapshot> e)
+                        {
+                            SubmitTreeUpdateAsync(
+                                (treeSnapshot, configuredProjectExports, ct) =>
+                                {
+                                    IProjectTree updatedTree = SyncNode(
+                                        imports: e.Value.Value,
+                                        tree: treeSnapshot.Value.Tree);
+
+                                    return Task.FromResult(new TreeUpdateResult(updatedTree, e.DataSourceVersions));
+                                });
+
+                            return;
+
+                            IProjectTree SyncNode(IReadOnlyList<IProjectImportSnapshot> imports, IProjectTree tree)
```

Comment:
4th level local function, a new record! 🎉 🏅

Reply:
In a property setter, no less.